src/System.Private.CoreLib/shared/System/Number.Parsing.cs

```diff
@@ -1468,64 +1468,87 @@ internal static decimal ParseDecimal(ReadOnlySpan<char> value, NumberStyles styl
 
         private static unsafe bool NumberBufferToDecimal(ref NumberBuffer number, ref decimal value)
         {
-            decimal d = new decimal();
-
             char* p = number.digits;
             int e = number.scale;
-            if (*p == 0)
+            bool sign = number.sign;
+            uint c = *p;
+            if (c == 0)
             {
                 // To avoid risking an app-compat issue with pre 4.5 (where some app was illegally using Reflection to examine the internal scale bits), we'll only force
                 // the scale to 0 if the scale was previously positive (previously, such cases were unparsable to a bug.)
-                if (e > 0)
+                value = new decimal(0, 0, 0, sign, (byte)Math.Clamp(-e, 0, 28));
+                return true;
+            }
+
+            if (e > DecimalPrecision)
+                return false;
+
+            ulong low64 = 0;
+            while (e > -28)
+            {
+                e--;
+                low64 *= 10;
+                low64 += c - '0';
+                c = *++p;
+                if (low64 >= ulong.MaxValue / 10)
+                    break;
+                if (c == 0)
                 {
-                    e = 0;
+                    while (e > 0)
+                    {
+                        e--;
+                        low64 *= 10;
+                        if (low64 >= ulong.MaxValue / 10)
+                            break;
+                    }
+                    break;
                 }
             }
-            else
+
+            uint high = 0;
+            while ((e > 0 || c != 0 && e > -28) &&
```

Comment:
I think it might be better to explicitly group the logical expressions when `||` and `&&` exist in the same expression (i.e. `e > 0 || (c != 0 && e > -28)`).